paritytech/capi
# feat: add devnet custom genesis, aura and grandpa support

Resolves #889 

**How to test?**

- for `aura` and `grandpa`, compile substrate `node-template`
- execute the following script
```typescript
import { bins, net } from "./mod.ts"

const controller = new AbortController()

Deno.addSignalListener("SIGTERM", () => controller.abort())
Deno.addSignalListener("SIGINT", () => controller.abort())

const bin = bins({
  nodeTemplate: "/Users/parity/ghq/github.com/paritytech/substrate/target/release/node-template",
})

const relay = net.dev({
  bin: bin.nodeTemplate,
  chain: "local",
  customize(chainSpec) {
    // @ts-ignore
    chainSpec.genesis.runtime.sudo.key = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"
  },
})
relay.name = "relay"

console.log(await relay.spawnNet(controller.signal, await Deno.makeTempDirSync({ dir: "./tmp" })))
```
- check the logs `./tmp/cbfda8c3/relay/alice/stderr` and validate that blocks are produced and finalized
- check the sudo key is `5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty` (or `await chain.Sudo.Key.value()`)
